File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.py
from collections import deque

MAX = 100_000

n, k = map(int, input().split())

visited = [False] * (MAX + 1)
q = deque()

q.append((n, 0))
visited[n] = True

while q:
    x, t = q.popleft()

    if x == k:
        print(t)
        break

    for nx in (x - 1, x + 1, 2 * x):
        if 0 <= nx <= MAX and not visited[nx]:
            visited[nx] = True
            q.append((nx, t + 1))

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.cc
#include <iostream>
#include <deque>
#include <vector>

#define MAX 100000

using namespace std;

struct Node {
    int x;
    int t;
};

int main() {
    int n, k;
    cin >> n >> k;
    
    deque<Node> q;
    vector<bool> visited(MAX + 1, false);
    q.push_back({n, 0});
    visited[n] = true;
    while(!q.empty()) {
        Node cur = q.front();
        q.pop_front();
        
        if (cur.x == k) {
            cout << cur.t;
            return 0;
        }
        if (cur.x + 1 <= MAX && !visited[cur.x + 1]) {
            q.push_back({cur.x + 1, cur.t + 1});
            visited[cur.x + 1] = true;
        }
        if (2 * cur.x <= MAX && !visited[2 * cur.x]) {
            q.push_back({2 * cur.x, cur.t + 1});
            visited[2 * cur.x] = true;
        }
        if (cur.x - 1 >= 0 && !visited[cur.x - 1]) {
            q.push_back({cur.x - 1, cur.t + 1});
            visited[cur.x - 1] = true;
        }
    }
}